### PR TITLE
Upgrade tool minor and patch commands

### PR DESCRIPTION
**Pull request by Bassel17**
### What does it do?

- Adds minor and patch upgrades to the upgrade tool

### How to test it?

- Make sure that running `minor` command upgrades to the latest minor of the current major, and executes codemods if any available.
- Make sure that running `patch` command upgrades to latest available patch of the current major and minor, and executes codemods if any available.

### CODE REVIEW

File: packages/utils/upgrade/src/modules/version/range.ts [hunk 1]
```diff
@@ -19,6 +22,14 @@ export const rangeFromReleaseType = (current: Version.SemVer, identifier: Versio
       const nextMajor = semver.inc(current, 'major') as Version.LiteralSemVer;
       return fromCurrentTo(nextMajor);
     }
+    case Version.ReleaseType.Patch: {
+      const preMinor = semver.inc(current, 'preminor') as Version.LiteralSemVer;
```
Comment (Convly) [inline packages/utils/upgrade/src/modules/version/range.ts:26]: Could you add a small comment with an example of what inc(version, 'preminor') does like what is done for major? (same for minor)
It could help with reading through the code later on

File: packages/utils/upgrade/src/modules/version/range.ts [hunk 2]
```diff
@@ -8,8 +8,11 @@ export const rangeFactory = (range: string): Version.Range => {
 };
 
 export const rangeFromReleaseType = (current: Version.SemVer, identifier: Version.ReleaseType) => {
-  const fromCurrentTo = (version: Version.LiteralVersion) => {
-    return rangeFactory(`>${current.raw} <=${version}`);
+  const fromCurrentTo = (version: Version.LiteralVersion, isUpperInclusive = true) => {
```
Comment (Convly) [inline packages/utils/upgrade/src/modules/version/range.ts:11]: Since the format isn't exactly the same everywhere, I feel like this helper introduces more complexity for a 3x one-liner, than it removes. What do you think about just using calls to `rangeFactory('>current < | <= version')`?

Comment (Convly) [reply packages/utils/upgrade/src/modules/version/range.ts:11]: (I know you just changed it already, sorry for coming in this late)

File: packages/utils/upgrade/src/modules/version/range.ts [hunk 3]
```diff
@@ -19,6 +19,14 @@ export const rangeFromReleaseType = (current: Version.SemVer, identifier: Versio
       const nextMajor = semver.inc(current, 'major') as Version.LiteralSemVer;
       return fromCurrentTo(nextMajor);
     }
+    case Version.ReleaseType.Patch: {
+      const preMinor = semver.inc(current, 'preminor') as Version.LiteralSemVer;
```
[the remark this answers is not in the document]
Comment (Convly) [reply packages/utils/upgrade/src/modules/version/range.ts:26]: Versions' tests are currently being developed by @christiancp100 in another branch
@christiancp100 , since your PR will be probably merged after this one, could you add a test for the minor with edge cases like Ben's? (when you'll have access to the actual code)

File: packages/utils/upgrade/src/tasks/upgrade/requirements/minor.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    const { project, npmVersionsMatches } = context;
+    if (npmVersionsMatches.length !== 1) {
+      const invalidVersions = npmVersionsMatches.slice(0, -1);
+      const invalidVersionsAsSemVer = invalidVersions
+        .filter((v) => semver.minor(v.version) === project.strapiVersion.minor)
+        .map((v) => v.version);
+      const nbInvalidVersions = invalidVersionsAsSemVer.length;
+      if (nbInvalidVersions !== 0) {
+        throw new Error(
+          `Doing a minor upgrade requires to be on the latest patch for v${project.strapiVersion} version, but found ${nbInvalidVersions} patch(es) ${invalidVersionsAsSemVer}`
```
Comment (Convly) [inline packages/utils/upgrade/src/tasks/upgrade/requirements/minor.ts:16]: Does it though? I believed you could just run a minor upgrade without having to run a patch and then a minor to accommodate both bumps directly

Also, I don't see this requirement in the feature specification, did I miss something regarding this?

**Review by Convly — COMMENTED** [review]
Also, it seems you will need to clean-up your git history, the commitlint job isn't happy (you can refer to `.commitlintrc.ts` for more info)

File: packages/utils/upgrade/src/tasks/upgrade/requirements/minor.ts (hunk 4 above)
Comment (Bassel17) [reply packages/utils/upgrade/src/tasks/upgrade/requirements/minor.ts:16]: Not sure maybe i understood the user story incorrectly

File: packages/utils/upgrade/src/modules/version/range.ts [hunk 5]
```diff
@@ -8,16 +8,24 @@ export const rangeFactory = (range: string): Version.Range => {
...
   switch (identifier) {
     case Version.ReleaseType.Major: {
       // semver.inc(_, 'major') will target <major + 1>.0.0 which is what we want
       // e.g. for 4.15.4, it'll return 5.0.0
       const nextMajor = semver.inc(current, 'major') as Version.LiteralSemVer;
-      return fromCurrentTo(nextMajor);
+      return rangeFactory(`>${current.raw} <=${nextMajor}`);
+    }
+    case Version.ReleaseType.Patch: {
+      // This will return the premionr for the next minor
```
Comment (Convly) [inline packages/utils/upgrade/src/modules/version/range.ts:19]:
```suggestion
      // This will return the preminor for the next minor
```

File: packages/utils/upgrade/src/modules/version/range.ts [hunk 6]
```diff
@@ -8,16 +8,24 @@ export const rangeFactory = (range: string): Version.Range => {
...
+      // e.g. for 4.15.4, it'll return 4.16.0-0
```
Comment (Convly) [inline packages/utils/upgrade/src/modules/version/range.ts:20]: That's annoying, the `Version.LiteralSemVer` tells us that the value is of type `<number>.<number>.<number>`, but we have a pre-release identifier in our variable.

I guess you could either remove the cast or modify the type to fit the value.

Comment (Convly) [reply packages/utils/upgrade/src/modules/version/range.ts:20]: On the same note, why are we using pre-releases instead of regular releases here? Is it to handle special cases?
Why couldn't we do inc('minor') instead?

Comment (Bassel17) [reply packages/utils/upgrade/src/modules/version/range.ts:20]: just guessing from my part but the the pre-releases would be part of the upgrade range, if we ever do a pre-release

Comment (Bassel17) [reply packages/utils/upgrade/src/modules/version/range.ts]: ok never mind it only includes pre-releases if we add it as an option

**Review by Convly — APPROVED** [review]
LGTM

Let's add something about writing tests for this in the up-coming sprint

**Comment by Convly** [comment]
Looks like tests were failing: https://github.com/strapi/strapi/actions/runs/7180787218/job/19556995396
